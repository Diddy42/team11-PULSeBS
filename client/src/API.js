async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/login.php', {     //Da verificare
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({username: username, password: password}),
        }).then((response) => {
            response.json().then((obj) => { 
                console.log(obj);
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
            })
            .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout(){
	return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
        .then((response) => {
			if(response.ok){
				resolve();
			}
			else{
				reject();
			}
       })
       .catch((err) => { 
      		 reject(err); 
       });
    });
}


async function getStudentLectures(studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const lectures = await response.json();
    if(response.ok) {
        return lectures;
    } else {
        let err = {status: response.status, errorObj: lectures};
        throw err; 
    }
}

async function getBooking(lectureId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const booking = await response.json();
    if(response.ok) {
        return booking;
    } else {
        let err = {status: response.status, errorObj: booking};
        throw err; 
    }
}

async function bookASeat(lectureId, studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({lectureId: lectureId, studentId: studentId}),
    });
    
    try{
        const resJ = await response.json();
        if(response.ok) {
		    return resJ;
		} else {
		    let err = {status: response.status, errorObj: resJ};
		    throw err; 
		}
    }
    catch(e){
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

const API = {
    userLogin, logout, getStudentLectures, getBooking, bookASeat
};
export default API;
