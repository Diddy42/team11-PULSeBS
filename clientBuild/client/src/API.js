async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost/server/src/api/Login.php', {
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({ username: username, password: password }),
        }).then((response) => {
            response.json().then((obj) => {
                /*
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
                */
                resolve(obj);
            })
                .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout() {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
            .then((response) => {
                if (response.ok) {
                    resolve(0);
                }
                else {
                    reject();
                }
            })
            .catch((err) => {
                reject(err);
            });
    });
}

//Ritorna le lezioni prenotabili dallo studente
async function getBookableStudentLectures(studentId) {
    const url = "http://localhost/server/bookableLessons/" + studentId;
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

//Ritorna le prenotazioni effettuate dagli studenti per la lezione indicata (teacher side)
async function getBooking(lectureId) {
    const url = "http://localhost/server/bookedStudentsForLecture/" + lectureId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

// API to retrieve all the bookings of a student
async function getStudentBookings(studentId) {
    const url = "http://localhost/server/studentBookings/" + studentId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

//API per prenotare un posto a lezione
async function bookASeat(lectureId, studentId, date) {
    const url = "http://localhost/server/insertLecture"
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
            idLesson: lectureId,
            idUser: studentId,
            date: date
        }),
    });

    try {
        const resJ = await response.json();
        if (response.ok) {
            return resJ;
        } else {
            let err = { status: response.status, errorObj: resJ };
            throw err;
        }
    }
    catch (e) {
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

// API to retrieve all the lectures of a teacher
async function getTeacherLectures(teacherId) {
    const url = "http://localhost/server/scheduledLecturesForTeacher/" + teacherId
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

async function deleteBooking(bookingId) {
    const url = "http://localhost/server/updateBooking/" + bookingId;
    return new Promise((resolve, reject) => {
        fetch(url, {
            method: 'PUT',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({
            }),
        }).then((response) => {
            if (response.ok) {
                resolve(0);
            } else {
                reject({ errore: "Error in deleting a booking" });
            }
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function deleteLecture(lectureId) {
    const url = "http://localhost/server/deleteLecture/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to delete the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function changeToOnline(lectureId) {
    const url = "http://localhost/server/changeToOnline/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to change the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function getAllCourses() {
    const url = "http://localhost/server/courses";

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const ALL_COURSES_FILTER = "L.idCourse";
async function getBookingStatisticsByMonth(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year,monthOfYear&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
} 
async function getBookingStatisticsByWeek(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year_month_week&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}
async function getBookingStatisticsByLesson(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=L.idLesson&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}

async function getCoursesOfTeacher(teacherId) {
    const url = "http://localhost/server/teacherCourses/" + teacherId;

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const API = {
    userLogin, logout, getBookableStudentLectures, getBooking, getStudentBookings, bookASeat, deleteBooking, getTeacherLectures, deleteLecture, changeToOnline, getAllCourses, 
    getBookingStatisticsByMonth, getBookingStatisticsByWeek, getBookingStatisticsByLesson, ALL_COURSES_FILTER, getCoursesOfTeacher
};
export default API;
