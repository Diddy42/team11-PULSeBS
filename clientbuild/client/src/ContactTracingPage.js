import React from 'react';
import { Container, Row, Col, Form, Alert, Button, Table } from 'react-bootstrap';
import API from './API';

class ContactTracingPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = { allStudents: [], filteredStudents: [], alwaysShowStudents: false, maxStudentsToShow: 100 };
    }

    render(){
        return <>
            <Container fluid>
                <Row>
                    <Col className='col-3'>
                        <Form className='mt-3'>
                            <Form.Group controlId="formStudentName">
                                <Form.Label>Enter the positive student's name:</Form.Label>
                                <Form.Control type="text" placeholder="Mario Rossi" data-testid="studentName-test" onChange={this.handleOnChangeText}/>
                            </Form.Group>
                        </Form>

                        <Alert variant={(this.state.filteredStudents.length > 0) ? 'success' : 'danger'}>
                            Number of students found: {this.state.filteredStudents.length}
                        </Alert>

                        { (this.state.filteredStudents.length > this.state.maxStudentsToShow && this.state.alwaysShowStudents === false) && <>
                          <Alert variant='danger'>
                        Too many students to show (> {this.state.maxStudentsToShow}). Please narrow your search first. <br></br>
                              <Button 
                                className='mt-3' 
                                variant='danger' 
                                onClick={() => {this.setState({alwaysShowStudents: true});}}>
                                    Show them anyway
                              </Button>
                          </Alert> 
                             
                        </> }

                        {  (this.state.filteredStudents.length > this.state.maxStudentsToShow && this.state.alwaysShowStudents === true) && 
                            <Alert variant='info'>
                                Currently showing {this.state.filteredStudents.length} students.
                                <Button 
                                className='mt-3' 
                                variant='success' 
                                onClick={() => {this.setState({alwaysShowStudents: false});}}>
                                    Restrict number of students to be shown
                            </Button>
                          </Alert>
                        }

                    </Col>

                    <Col className='col-8'>
                        <Table className='mt-3' striped bordered hover>
                            <thead>
                                <tr>
                                    <th>Name</th>
                                    <th>Email</th>
                                    <th>Student id</th>
                                    <th>Action</th>
                                </tr>
                            </thead>

                            <tbody>
                                {(this.state.filteredStudents.length <= this.state.maxStudentsToShow || this.state.alwaysShowStudents === true) &&
                                    this.state.filteredStudents.map(s => this.createStudentTableRow(s))}
                            </tbody>
                        </Table>
                    </Col>
                </Row>
            </Container>
        </>
    }

    componentDidMount = () => {
        API.getAllStudents()
        .then((res) => {
            this.setState({
                allStudents: res
            });
        });
    }

    handleStudentButtonClick = (s) => {
        console.log('clicked student: ');
        console.log(s);
    }

    createStudentTableRow = (s) => {
        return <tr>
            <td>{s.name}</td>
            <td>{s.email}</td>
            <td>{s.idStudent}</td>
            <td>
                <Button className='mt-1' variant='danger'data-testid="button-test" onClick={() => {this.handleStudentButtonClick(s);}}>Generate report from this student</Button>
            </td>
        </tr>
    }

    studentMatchesInput = (s, wordsInInput) => {
        var res = true;
        var i;

        for(i = 0; i < wordsInInput.length; i++){
            if(!s.name.toLowerCase().includes(wordsInInput[i])){
                res = false;
            }
        }

        return res;
    }

    handleOnChangeText = (event) => {
        var currentTextInput = event.target.value.toLowerCase();
        var wordsInInput = currentTextInput.split(' ').filter(w => w !== '');
        
        var filteredStudents = this.state.allStudents.filter(s => this.studentMatchesInput(s, wordsInInput));
        this.setState({ filteredStudents: filteredStudents });
    }
}

export default ContactTracingPage;