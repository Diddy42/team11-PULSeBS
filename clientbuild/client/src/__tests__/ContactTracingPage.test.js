import TestRenderer from 'react-test-renderer';
import React from 'react';
import ContactTracingPage from '../ContactTracingPage';
import API from '../API';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { fireEvent } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test('renders learn react link', () => {
    const tree = TestRenderer.create(<ContactTracingPage/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  
test('student name input box', () => {
  API.getAllStudents = jest.fn(() => Promise.resolve([{name: "test1", userId: 1}, {name: "test2", userId: 2}, {name: "student3", userId: 3}]));
  const {queryByTestId} = render(<ContactTracingPage/>);
  const studentNameInput = queryByTestId("studentName-test");
  fireEvent.change(studentNameInput, {target: {value: "test1"}});
  expect(studentNameInput.value).toBe("test1");
});

test("studentMatchesInput", () => {
  API.getAllStudents = jest.fn(() => Promise.resolve([{name: "test1", userId: 1}, {name: "test2", userId: 2}, {name: "student3", userId: 3}]));
  const component = shallow(<ContactTracingPage/>);
  const instance = component.instance();
  const res1 = instance.studentMatchesInput({name: "test1"}, "pos");
  expect(res1).toBe(false);
  const res2 = instance.studentMatchesInput({name: "test1"}, "tes");
  expect(res2).toBe(true);
});

test("createStudentTableRow", () => {
  API.getAllStudents = jest.fn(() => Promise.resolve([{name: "test1", userId: 1}, {name: "test2", userId: 2}, {name: "student3", userId: 3}]));
  const component = shallow(<ContactTracingPage/>);
  const instance = component.instance();
  const res = instance.createStudentTableRow({name: "testName", email: "testEmail", idStudent: "testID"});
  console.log(res);
  expect(res).not.toBe(undefined);
});

/*
test('updateCourses', async () => {
  const component = shallow(<TeacherHistoricalDataPage teacherId={2}/>);
  const instance = component.instance();
  API.getCoursesOfTeacher = jest.fn(() => Promise.resolve([{courseName: "test"}]));
  await instance.updateCourses();
  expect(instance.state.currentCourse).toBe("test");
});

test('getDataFromServerAndUpdate', async () => {
  const component = shallow(<TeacherHistoricalDataPage teacherId={2}/>);
  const instance = component.instance();
  API.getTeacherStatistics = jest.fn(() => Promise.resolve([{numberOfBookings: 10, dataLecture: 1}, {numberOfBookings: 20, dataLecture: 2}]));
  instance.updateGraphData = jest.fn();
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Lecture"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  await instance.getDataFromServerAndUpdate();
  expect(instance.updateGraphData).toHaveBeenCalledTimes(2);
  API.getTeacherStatistics = jest.fn(() => Promise.resolve([{average: 1, weekOfYear: 1, year: 1}, {average: 2, weekOfYear: 2, year: 2}]));
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Week"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  await instance.getDataFromServerAndUpdate();
  expect(instance.updateGraphData).toHaveBeenCalledTimes(3);
  API.getTeacherStatistics = jest.fn(() => Promise.resolve([{average: 1, monthOfYear: 1, year: 1}, {average: 2, monthOfYear: 2, year: 2}]));
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Month"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  await instance.getDataFromServerAndUpdate();
  expect(instance.updateGraphData).toHaveBeenCalledTimes(4);
});

test('getDataFromServerAndUpdate API error', async () => {
  const component = shallow(<TeacherHistoricalDataPage teacherId={2}/>);
  const instance = component.instance();
  API.getTeacherStatistics = jest.fn(() => Promise.reject([]));
  instance.updateGraphData = jest.fn();
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Lecture"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  try {
    await instance.getDataFromServerAndUpdate();
    expect(true).toBe(false);
  } catch(e) {
      expect(true).toBe(true);
  }
  //expect(instance.updateGraphData).toHaveBeenCalledTimes(2);
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Week"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  try {
    await instance.getDataFromServerAndUpdate();
    expect(true).toBe(false);
  } catch(e) {
      expect(true).toBe(true);
  }
  await instance.setState({currentCourse: "test"});
  await instance.setState({currentDetail: "Month"});
  await instance.setState({courses: [{courseName: "test", idCourse: 1}]});
  try {
    await instance.getDataFromServerAndUpdate();
    expect(true).toBe(false);
  } catch(e) {
      expect(true).toBe(true);
  }
});

test('handleNewDetailClick and handleNewCourseClick', async () => {
  const component = shallow(<TeacherHistoricalDataPage teacherId={2}/>);
  const instance = component.instance();
  const eventMock = {
    currentTarget: {
      innerText: "test",
    }
  };
  instance.getDataFromServerAndUpdate = jest.fn();
  await instance.handleNewDetailClick(eventMock);
  expect(instance.state.currentDetail).toBe("test");
  expect(instance.getDataFromServerAndUpdate).toHaveBeenCalledTimes(2);
  await instance.handleNewCourseClick(eventMock);
  expect(instance.state.currentDetail).toBe("test");
  expect(instance.getDataFromServerAndUpdate).toHaveBeenCalledTimes(3);
});
*/