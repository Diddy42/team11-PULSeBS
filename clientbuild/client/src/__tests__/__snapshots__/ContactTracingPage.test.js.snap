// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-3 col"
    >
      <form
        className="mt-3"
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="formStudentName"
          >
            Enter the positive student's name:
          </label>
          <input
            className="form-control"
            data-testid="studentName-test"
            id="formStudentName"
            onChange={[Function]}
            placeholder="Mario Rossi"
            type="text"
          />
        </div>
      </form>
      <div
        className="fade alert alert-danger show"
        role="alert"
      >
        Number of students found: 
        0
      </div>
    </div>
    <div
      className="col-8 col"
    >
      <table
        className="mt-3 table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Email
            </th>
            <th>
              Student id
            </th>
            <th>
              Action
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
